### PR TITLE
Day17 is finally here

### PR DESCRIPTION
**Pull request by mtsochacki**

### CODE REVIEW

File: 2021/day17/day17.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            }
+        }
+        int total = 0;
+        for (Integer xVelocity : xVelocities) {
+            for (Integer yVelocity : yVelocities) {
+                if (isInTrench(xVelocity, yVelocity, xStart, xEnd, yStart, yEnd)) {
+                    total++;
+                }
+            }
+        }
```
Comment (Lammatian) [inline 2021/day17/day17.java:99]: I guess this is the reason why you have all the functions above even though they're basically all the same logic with minor differences. Maybe you could write just one function with additional arguments `boolean checkX, boolean checkY` that would do the calculations only for the axes that have this `boolean`s set to `true`? I don't exactly know how to make this work and be nice, but I just don't like the repetition.

**Review by Lammatian — CHANGES_REQUESTED** [review]
This one's tough - a lot of the code repeats itself, but on the other hand I don't see one best way to make it DRY. I mentioned one possibility in one comment, another would maybe be a `Simulator` class with a method `simulate(Tuple2D point, Tuple2D velocity, borders2D)` with two subclasses: `YSimulator` and `XSimulator` that would have their own versions of `simulate(Tuple1D point, Tuple1D velocity, borders1D)`. `Tuple2D` and `Tuple1D` could be implementations of a `Tuple` base class.

File: 2021/day17/day17.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        return true;
+    }
+
+    public static int[] readInput(String filename){
+        int[] trenchCoordinates = new int[4];
+        try {
+            Scanner sc = new Scanner(new File(filename));
+            String[] newLine = sc.nextLine().split(":")[1].split(",");
+            String[] coordinatesX = newLine[0].split("=")[1].split("\\..");
+            String[] coordinatesY = newLine[1].split("=")[1].split("\\..");
```
[the remark this answers is not in the document]
Comment (mtsochacki) [reply 2021/day17/day17.java:65]: I WOULD BE IN TROUBLE

File: 2021/day17/day17.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import java.io.File;
+
+public class day17 {
+    public static boolean isYInTrench(int yVelocity, int yStart, int yEnd) {
+        int y = 0;
+        while (!(y >= yEnd && y <= yStart)) {
+            y += yVelocity;
+            if (y < yEnd) {
+                return false;
+            }
```
[the remark this answers is not in the document]
Comment (mtsochacki) [reply 2021/day17/day17.java:12]: It was really weird. Original version was even more weird.
I rewrote it to 5 lines of not weird methods now :D

**Comment by mtsochacki** [comment]
I wasn't able to resolve everything, but I will merge it for now and have another look later

**Pull request merged**